File: example/src/test/scala/ru/tinkoff/load/example/feeders/Feeders.scala
package ru.tinkoff.load.example.feeders

import io.gatling.core.Predef._
import io.gatling.core.feeder._
import ru.tinkoff.gatling.feeders._
import ru.tinkoff.gatling.utils.RandomDataGenerators
import ru.tinkoff.gatling.utils.phone._

import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.time.{LocalDateTime, ZoneId}

object Feeders {

  private val newYearDate                       = LocalDateTime.of(2020, 1, 1, 0, 0)
  private val goToWorkDate                      = LocalDateTime.of(2020, 1, 9, 9, 0)
  private val formatterShort: DateTimeFormatter = DateTimeFormatter.ofPattern("MM:dd")

//  override implicit def configuration: GatlingConfiguration = GatlingConfiguration.loadForTest()

  // date2pattern
  val timeShort: Feeder[String] = CurrentDateFeeder("timeShort", formatterShort)

  // random date +/- 5 minutes with "Australia/Sydney" timezone
  val ausTZ: ZoneId                  = ZoneId.of("Australia/Sydney")
  val timezoneRandom: Feeder[String] =
    RandomDateFeeder("timezoneRandom", 5, 5, "hh:mm:dd", unit = ChronoUnit.MINUTES, timezone = ausTZ)

  // random date +/- 3 days from now
  val simpleRandomDate: Feeder[String] = RandomDateFeeder("simpleDate", 3, 3)

  // random date from newYearDate  with specified date string pattern
  val holidaysDate: Feeder[String] = RandomDateFeeder("holidays", 8, 0, "yyyy-MM-dd'T'HH:mm", newYearDate, ChronoUnit.DAYS)

  // random time from 9:00 to 18:00
  val firstWorkDayHours: Feeder[String] =
    RandomDateFeeder("firstWorkDayHours", 9 * 60, 0, "HH:mm", goToWorkDate, ChronoUnit.MINUTES)

  // feeder provide two params:
  // startOfVacation = LocalDateTime.now()
  // endOfVacation = random date from now() to 14 days in the future
  val vacationDate: Feeder[String] =
    RandomDateRangeFeeder("startOfVacation", "endOfVacation", 14, "yyyy-MM-dd", LocalDateTime.now(), ChronoUnit.DAYS)

  // random Int
  val randomDigit: Feeder[Int]      = RandomDigitFeeder("randomDigit")
  val randomRangeInt: Feeder[Int]   = CustomFeeder[Int]("randomRangeInt", RandomDataGenerators.randomDigit(1, 50))
  val randomRangeFloat: Feeder[Any] =
    CustomFeeder("randomRangeFloat", RandomDataGenerators.randomDigit(1.toFloat, 10.toFloat))

  // random phone
  // +7 country code is default
  val randomPhone: Feeder[String] = RandomPhoneFeeder("randomPhone")

  // random USA phone
  val usaPhoneFormats: PhoneFormat =
    PhoneFormat(
      countryCode = "+1",
      length = 10,
      areaCodes = Seq("484", "585", "610"),
      prefixes = Seq("55", "81", "111"),
      format = "+X XXX XXX-XX-XX",
    )

  val randomUsaPhone: Feeder[String] = RandomPhoneFeeder("randomUsaPhone", usaPhoneFormats)

  val phoneFormatsFromFile: String   = "phoneTemplates/ru.json"
  val ruMobileFormat: PhoneFormat    = PhoneFormat(
    countryCode = "+7",
    length = 10,
    areaCodes = Seq("903", "906", "908"),
    prefixes = Seq("55", "81", "111"),
    format = "+X XXX XXX-XX-XX",
  )
  val ruCityPhoneFormat: PhoneFormat = PhoneFormat(
    countryCode = "8",
    length = 7,
    areaCodes = Seq("495", "495"),
    format = "X(XXX)XXX-XX-XX",
  )
  val ruPhoneFormats                 = List(ruMobileFormat, ruCityPhoneFormat)

  val simplePhoneNumber: Feeder[String]                 = RandomPhoneFeeder("simplePhoneFeeder")
  val randomPhoneNumberFromJson: Feeder[String]         =
    RandomPhoneFeeder("randomPhoneNumberFile", phoneFormatsFromFile)
  val randomPhoneNumber: Feeder[String]                 =
    RandomPhoneFeeder("randomPhoneNumber", ruPhoneFormats: _*)
  val randomE164PhoneNumberFromJson: Feeder[String]     =
    RandomPhoneFeeder("randomE164PhoneNumberFile", phoneFormatsFromFile, TypePhone.E164PhoneNumber)
  val randomE164PhoneNumber: Feeder[String]             =
    RandomPhoneFeeder("randomE164PhoneNumber", TypePhone.E164PhoneNumber, ruMobileFormat, ruCityPhoneFormat)
  val randomTollFreePhoneNumberFromJson: Feeder[String] =
    RandomPhoneFeeder("randomTollFreePhoneNumberFile", phoneFormatsFromFile, TypePhone.TollFreePhoneNumber)
  val randomTollFreePhoneNumber: Feeder[String]         =
    RandomPhoneFeeder("randomTollFreePhoneNumber", TypePhone.TollFreePhoneNumber, ruMobileFormat)

  // random alphanumeric String with specified length
  val randomString: Feeder[String] = RandomStringFeeder("randomString", 16)

  // random String generated from specified alphabet (or alphanumeric as default)
  // with random length in specified interval from 1 to 10
  val randomRangeString: Feeder[String] =
    RandomRangeStringFeeder("randomRangeString", 1, 10, "qwertyuiop*+-123")

  // random UUID
  val randomUuid: Feeder[String] = RandomUUIDFeeder("randomUuid")

  // sequence of Long numbers from one to Long.MaxValue with specified step = 2
  val sequenceLong: Feeder[Long] = SequentialFeeder("sequenceLong", 1, 2)

  private def myFunction: String = {
    scala.util.Random.shuffle(List("first", "second", "third")).head
  }

  // custom feeder from provided function
  val myCustomFeeder: Feeder[String] = CustomFeeder[String]("myParam", myFunction)

  private val digitFeeder  = RandomDigitFeeder("digit")
  private val stringFeeder = RandomStringFeeder("string")
  private val phoneFeeder  = RandomStringFeeder("phone")

  // Vault HC feeder
  private val vaultUrl                       = System.getenv("vaultUrl")
  private val secretPath                     = System.getenv("secretPath")
  private val roleId                         = System.getenv("roleId")
  private val secretId                       = System.getenv("secretId")
  private val keys                           = List("k1", "k2", "k3")
  val vaultFeeder: FeederBuilderBase[String] = VaultFeeder(vaultUrl, secretPath, roleId, secretId, keys)

  // Get separated values feeder from the source
  // SeparatedValuesFeeder will return Vector(Map(HOSTS -> host11), Map(HOSTS -> host12), Map(USERS -> user11), Map(HOSTS -> host21), Map(HOSTS -> host22), Map(USERS -> user21), Map(USERS -> user22), Map(USERS -> user23))
  val vaultData: FeederBuilderBase[String]             = Vector(
    Map(
      "HOSTS" -> "host11,host12",
      "USERS" -> "user11",
    ),
    Map(
      "HOSTS" -> "host21,host22",
      "USERS" -> "user21,user22,user23",
    ),
  )
  val separatedValuesFeeder: FeederBuilderBase[String] =
    SeparatedValuesFeeder(None, vaultData.readRecords, ',')

  // how to combine together 2 or more feeders
  // as result we get feeder with 3 params: digit, string, phone
  val gluedTogetherFeeder: Feeder[Any] = digitFeeder ** stringFeeder ** phoneFeeder

  // transform values of this Feeder
  val finiteRandomDigitsWithTransform: FeederBuilderBase[Any] = RandomDigitFeeder("randomDigit")
    .toFiniteLength(20)
    .transform { case (k, v) => k -> v.toString }
    .circular

  // transform List to Feeder
  val list2feeder: FeederBuilderBase[Int] = List(1, 2, 3).toFeeder("listId").circular

  // string sequentially generated from the specified pattern
  val regexString: Feeder[String] = RegexFeeder("regex", "[a-zA-Z0-9]{8}")

  // random PAN
  val feederWithoutBinPAN: Feeder[String] = RandomPANFeeder("feederWithoutBinPAN")
  val feederPAN: Feeder[String]           = RandomPANFeeder("feederPAN", "421345", "541673")

  // random ITN
  val feederNatITN: Feeder[String] = RandomNatITNFeeder("feederNatITN")
  val feederJurITN: Feeder[String] = RandomJurITNFeeder("feederJurITN")

  // random OGRN
  val feederOGRN: Feeder[String] = RandomOGRNFeeder("feederOGRN")

  // random PSRNSP
  val feederPSRNSP: Feeder[String] = RandomPSRNSPFeeder("feederPSRNSP")

  // random KPP
  val feederKPP: Feeder[String] = RandomKPPFeeder("feederKPP")

  // random SNILS
  val feederSNILS: Feeder[String] = RandomSNILSFeeder("randomSNILS")

  // random russian passport
  val feederRusPassport: Feeder[String] = RandomRusPassportFeeder("feederRusPassport")

}

File: src/main/java/ru/tinkoff/gatling/javaapi/Feeders.java
package ru.tinkoff.gatling.javaapi;

import java.time.temporal.ChronoUnit;
import java.util.*;
import java.time.*;
import java.time.temporal.TemporalUnit;
import java.time.format.DateTimeFormatter;

import static scala.jdk.javaapi.CollectionConverters.asScala;

import ru.tinkoff.gatling.utils.phone.PhoneFormat;
import ru.tinkoff.gatling.utils.phone.TypePhone;
import scala.Function0;
import scala.Option;

import static ru.tinkoff.gatling.javaapi.internal.Feeders.*;

public final class Feeders {

    private Feeders() {
    }

    public static Iterator<Map<String, Object>> CurrentDateFeeder(String paramName, DateTimeFormatter datePattern, ZoneId timezone) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.CurrentDateFeeder.apply(paramName, datePattern, timezone));
    }

    public static Iterator<Map<String, Object>> CurrentDateFeeder(String paramName, DateTimeFormatter datePattern) {
        return CurrentDateFeeder(paramName, datePattern, ZoneId.systemDefault());
    }

    public static <T> Iterator<Map<String, Object>> CustomFeeder(String paramName, Function0<T> f) {
        return toJavaFeeder(
                ru.tinkoff.gatling.feeders.CustomFeeder.apply(paramName, f)
        );
    }

    public static Iterator<Map<String, Object>> RandomDateFeeder(
            String paramName,
            Integer positiveDaysDelta,
            Integer negativeDaysDelta,
            String datePattern,
            LocalDateTime dateFrom,
            TemporalUnit unit,
            ZoneId timezone
    ) {
        return toJavaFeeder(
                ru.tinkoff.gatling.feeders.RandomDateFeeder.apply(
                        paramName,
                        positiveDaysDelta,
                        negativeDaysDelta,
                        datePattern,
                        dateFrom,
                        unit,
                        timezone
                )
        );
    }

    public static Iterator<Map<String, Object>> RandomDateFeeder(
            String paramName,
            Integer positiveDaysDelta,
            Integer negativeDaysDelta
    ) {
        return RandomDateFeeder(
                paramName,
                positiveDaysDelta,
                negativeDaysDelta,
                "yyyy-MM-dd",
                LocalDateTime.now(),
                ChronoUnit.DAYS,
                ZoneId.systemDefault()
        );
    }

    public static Iterator<Map<String, Object>> RandomDateRangeFeeder(
            String paramNameFrom,
            String paramNameTo,
            Long offsetDate,
            String datePattern,
            LocalDateTime dateFrom,
            TemporalUnit unit,
            ZoneId timezone
    ) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomDateRangeFeeder.apply(paramNameFrom, paramNameTo, offsetDate, datePattern, dateFrom, unit, timezone));
    }

    public static Iterator<Map<String, Object>> RandomDateRangeFeeder(
            String paramNameFrom,
            String paramNameTo,
            Long offsetDate
    ) {
        return RandomDateRangeFeeder(paramNameFrom, paramNameTo, offsetDate, "yyyy-MM-dd", LocalDateTime.now(), ChronoUnit.DAYS, ZoneId.systemDefault());
    }

    public static Iterator<Map<String, Object>> RandomDigitFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomDigitFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomJurITNFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomJurITNFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomKPPFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomKPPFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomNatITNFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomNatITNFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomOGRNFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomOGRNFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomPANFeeder(String paramName, String... bins) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPANFeeder.apply(paramName, asScala(Arrays.asList(bins)).toSeq()));
    }

    public static Iterator<Map<String, Object>> RandomPhoneFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPhoneFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomPhoneFeeder(String paramName, PhoneFormat... formats) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPhoneFeeder.apply(paramName, asScala(Arrays.asList(formats)).toSeq()));
    }

    public static Iterator<Map<String, Object>> RandomPhoneFeeder(String paramName, TypePhone.TypePhone typePhone, PhoneFormat... formats) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPhoneFeeder.apply(paramName, typePhone, asScala(Arrays.asList(formats)).toSeq()));
    }

    public static Iterator<Map<String, Object>> RandomPhoneFeeder(String paramName, String formatsPath, TypePhone.TypePhone typePhone) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPhoneFeeder.apply(paramName, formatsPath, typePhone));
    }

    public static Iterator<Map<String, Object>> RandomPhoneFeeder(String paramName, String formatsPath) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPhoneFeeder.apply(paramName, formatsPath));
    }

    public static Iterator<Map<String, Object>> RandomPSRNSPFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomPSRNSPFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomRangeStringFeeder(String paramName, Integer from, Integer to, String alphabet) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomRangeStringFeeder.apply(paramName, from, to, alphabet));
    }

    public static Iterator<Map<String, Object>> RandomRangeStringFeeder(String paramName) {
        String alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789@#%\"&*()_-+={}<>?|:[].~";
        return RandomRangeStringFeeder(paramName, 10, 15, alphabet);
    }

    public static Iterator<Map<String, Object>> RandomRusPassportFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomRusPassportFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomSNILSFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomSNILSFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RandomStringFeeder(String paramName, Integer paramLength) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomStringFeeder.apply(paramName, paramLength));
    }

    public static Iterator<Map<String, Object>> RandomStringFeeder(String paramName) {
        return RandomStringFeeder(paramName, 10);
    }

    public static Iterator<Map<String, Object>> RandomUUIDFeeder(String paramName) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RandomUUIDFeeder.apply(paramName));
    }

    public static Iterator<Map<String, Object>> RegexFeeder(String paramName, String regex) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.RegexFeeder.apply(paramName, regex));
    }

    public static final class SeparatedValuesFeeder {
        private static final char commaSeparator = ',';
        private static final char SemicolonSeparator = ';';
        private static final char TabulationSeparator = '\t';

        public static Iterator<Map<String, Object>> apply(String paramName, String source, char separator) {
            return toJavaFeeder(ru.tinkoff.gatling.feeders.SeparatedValuesFeeder.apply(paramName, source, separator));
        }

        public static Iterator<Map<String, Object>> apply(String paramName, List<String> source, char separator) {
            return toJavaFeeder(ru.tinkoff.gatling.feeders.SeparatedValuesFeeder.apply(paramName, asScala(source).toSeq(), separator, io.gatling.core.Predef.configuration()));
        }

        public static Iterator<Map<String, Object>> apply(Optional<String> paramPrefix, List<Map<String, Object>> source, char separator) {
            Option<String> _paramPrefix = toScalaOption(paramPrefix);

            return toJavaFeeder(ru.tinkoff.gatling.feeders.SeparatedValuesFeeder.apply(_paramPrefix, toScala(source), separator, io.gatling.core.Predef.configuration()));
        }

        public static Iterator<Map<String, Object>> csv(String paramName, String source) {
            return apply(paramName, source, commaSeparator);
        }

        public static Iterator<Map<String, Object>> ssv(String paramName, String source) {
            return apply(paramName, source, SemicolonSeparator);
        }

        public static Iterator<Map<String, Object>> tsv(String paramName, String source) {
            return apply(paramName, source, TabulationSeparator);
        }

        public static Iterator<Map<String, Object>> csv(String paramName, List<String> source) {
            return apply(paramName, source, commaSeparator);
        }

        public static Iterator<Map<String, Object>> ssv(String paramName, List<String> source) {
            return apply(paramName, source, SemicolonSeparator);
        }

        public static Iterator<Map<String, Object>> tsv(String paramName, List<String> source) {
            return apply(paramName, source, TabulationSeparator);
        }

        public static Iterator<Map<String, Object>> csv(Optional<String> paramPrefix, List<Map<String, Object>> source) {
            return apply(paramPrefix, source, commaSeparator);
        }

        public static Iterator<Map<String, Object>> ssv(Optional<String> paramPrefix, List<Map<String, Object>> source) {
            return apply(paramPrefix, source, SemicolonSeparator);
        }

        public static Iterator<Map<String, Object>> tsv(Optional<String> paramPrefix, List<Map<String, Object>> source) {
            return apply(paramPrefix, source, TabulationSeparator);
        }

    }

    public static Iterator<Map<String, Object>> SequentialFeeder(String paramName, Integer start, Integer step) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.SequentialFeeder.apply(paramName, start, step));
    }

    public static Iterator<Map<String, Object>> SequentialFeeder(String paramName) {
        return SequentialFeeder(paramName, 0, 1);
    }

    public static Iterator<Map<String, Object>> VaultFeeder(
            String vaultUrl,
            String secretPath,
            String roleId,
            String secretId,
            List<String> keys
    ) {
        return toJavaFeeder(ru.tinkoff.gatling.feeders.VaultFeeder.apply(vaultUrl, secretPath, roleId, secretId, asScala(keys).toList()));
    }
}
